command/filemanager.go

```diff
@@ -53,3 +60,133 @@ func (fm *FileManager) CopyInlineFile(name string, fileContent pulumi.StringInpu
+}
+
+// CopyRelativeFile copies relative path to a remote path.
+// The relative path is defined in the same way as for `CopyRelativeFolder`.
+// This function returns the resource that can be used with `pulumi.DependsOn`.
+func (fm *FileManager) CopyRelativeFile(relativePath string, remotePath string, opts ...pulumi.ResourceOption) (pulumi.Resource, error) {
+	fullPath, _, err := getFullPath(relativePath, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	return fm.CopyFile(fullPath, remotePath, opts...)
+}
+
+// CopyFSFolder copies recursively a local folder to a remote folder.
+// You may consider using `CopyRelativeFolder` which has a simpler API.
+func (fm *FileManager) CopyFSFolder(
+	resourceName string,
+	folderFS fs.FS,
+	folderPath string,
+	remoteFolder string,
+	opts ...pulumi.ResourceOption) ([]pulumi.Resource, error) {
+	useSudo := true
+	folderCommand, err := fm.CreateDirectory(resourceName, pulumi.String(remoteFolder), useSudo, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create a temporary folder: %v for resource %v", err, resourceName)
+	}
+
+	files, folders, err := readFilesAndFolders(folderFS, folderPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read files and folders for %v", folderPath)
+	}
+
+	var folderResources []pulumi.Resource
+	for _, folder := range folders {
+		destFolder, err := getDestinationPath(folder, folderPath)
+		if err != nil {
+			return nil, err
+		}
+		remotePath := path.Join(remoteFolder, destFolder)
+		resources, err := fm.CreateDirectory("createFolder-"+remotePath, pulumi.String(remotePath), useSudo, pulumi.DependsOn([]pulumi.Resource{folderCommand}))
+		if err != nil {
+			return nil, err
+		}
+		folderResources = append(folderResources, resources)
+	}
+
+	fileResources := []pulumi.Resource{}
+	for _, file := range files {
+		destFile, err := getDestinationPath(file, folderPath)
+
+		if err != nil {
+			return nil, err
+		}
+
+		fileContent, err := fs.ReadFile(folderFS, file)
+		if err != nil {
+			return nil, err
+		}
+		fileCommand, err := fm.CopyInlineFile(
+			resourceName+"-"+file,
+			pulumi.String(fileContent),
+			path.Join(remoteFolder, destFile),
+			useSudo,
+			pulumi.DependsOn(folderResources))
+		if err != nil {
+			return nil, err
+		}
+		fileResources = append(fileResources, fileCommand)
+	}
+
+	return fileResources, nil
+}
+
+// When copying foo/bar to /tmp the result folder is /tmp/bar
+// This function remove the root prefix from the path (`foo` in this case)
+func getDestinationPath(folder string, rootFolder string) (string, error) {
+	baseFolder := filepath.Base(rootFolder)
+	rootWithoutBase := rootFolder[:len(rootFolder)-len(baseFolder)]
+
+	if !strings.HasPrefix(folder, rootWithoutBase) {
+		return "", fmt.Errorf("%v doesn't have the prefix %v", folder, rootWithoutBase)
+	}
+
+	return folder[len(rootWithoutBase):], nil
+}
+
+func getFullPath(relativeFolder string, skip int) (string, string, error) {
+	_, file, _, ok := runtime.Caller(skip)
```

Comment:
If I'm not mistaken `runtime.Caller` returns the build path, which will be different depending on many things (if it's built in `test-infra` or `datadog-agent`, another machine, CI, etc.).

By design, the equivalent of `__file__` doest not exist in `Go`, so I'm not sure this code is entirely reliable. That's the reason why the existing API takes full path or embed strings.

Reply:
`runtime.Caller` returns the **full path**. It makes no difference, if it's built in `test-infra` or `datadog-agent` (it works the same in both cases). If the binary is moved to another machine then copying the data makes a lot of sense.

Reply:
It makes a lot of sense in case of remote execution, artifact copy in CI, docker mounts, etc. IMHO if the caller wants to do that, it can call `runtime.Caller()` itself. It would make this API reliable in any case and defer the decision to the one that really knows where the file is.

Reply:
I understand the use cases you described but I think they doesn't apply for what we are doing:
- For running E2E tests, `go test` requires the source code
- For creating an environment, `pulumi up` requires also the source code
The only case where it can theoretically happen is when creating an environment from `datadog-agent` repository and the creation of the environment is done using a binary that is moved to another host. It is unlikely to happen and the user can still use embedded file in this very particular case.
I am a little hesitant to remove a feature that make life easier for the users because there is a theoretical case where it doesn't work. @vboulineau  What do you think?

Reply:
In which case `CopyFolder(fullPath string...)` can be used as we don't know the full path (We don't know where the code of `datadog-agent` is)

Reply:
I don't think there's a single use case of copying things from `datadog-agent` repo (even though, we have several ways of getting the repo path (through `git` for instance), it's especially easy in the CI), we may want to copy generated content (from `/tmp`) or downloaded content.

Reply:
Function `CopyAbsoluteFolder` added.